Development

Sly-Agustin opened the pull request.

Sly-Agustin committed:
Implented health endpoint

- Implemented health endpoint
- created a controller
- Added routes
- created server.js
- Added packages to package.json for execution
- updated gitignore

Sly-Agustin committed:
Implemented basic create

- Implemented basic dream creation, there is no validation ATM, but inserts in online MongoDB fine if parameters are passed correctly.
- Added controller for dream creation
- Modified routes to include the new method

Sly-Agustin committed:
Update healthController.js

- Now health route succeed if the controller connects with the DB. It also send an error 500 if it fails to connect.

Sly-Agustin committed:
Added validation to schema

Sly-Agustin committed:
Update dream.js

- Added some specific validation fields

Sly-Agustin committed:
Validation

- Added validation rules on schema
- Added validation on the controller. It handles invalid input for all fields

Sly-Agustin committed:
Update createQuestionnaireController.js

Deleted comments

Sly-Agustin committed:
Added students

- Added students, they have a studentId that is autoincremented in MongoDB.
- Added a controller, router, and routes for student creation
- Tweaks on server.js for student

Sly-Agustin committed:
Updated controller

- Updated the getDreamControllerById, now it can fetch existing questionnaires. Validation still not implented for this.
- Renamed routes

Sly-Agustin committed:
Refactor

- Replaced a lot of "dreams" with "questionnaire"

Sly-Agustin committed:
Refactor createQuestionnaire

- Refactor createQuestionnaireController.js

Sly-Agustin committed:
Refactor

- Reimplemented getDreamControllerById
- The app connects to the server as soon as it start and not everytime it does a request
- Deleted console logs

Sly-Agustin committed:
Test getById

- Added mock test for getById
- Removed exec() from getById controller

Sly-Agustin committed:
Added test for create a questionnaire

mjuan99 commented on src/controller/getDreamControllerById.js:
```diff
@@ -0,0 +1,47 @@
+const mongoose = require('mongoose');
+
+function isNumber(posibleId){
+	let id = parseInt(posibleId, 10);
```
You are using parseInt two times over the same id (once in idVerification() and again here)

mjuan99 commented on src/controller/getDreamControllerById.js:
```diff
@@ -0,0 +1,47 @@
+
+function idVerification(possibleId) {
+    const id = parseInt(possibleId, 10);
+    if (!isNumber(id)){
```
I think you could replace `!isNumber(id)` with `isNaN(id)` and remove the isNumber() function

Sly-Agustin replied:
After some research i found that both methods can cause errors because of the way they are implemented. 
`parseInt()` can parse invalid strings in numbers, for example: `parseInt("56ex")` is converted into `56` which should not occur. 
`isNaN()` can fail in some cases where javascript behaves strangely, for example: `isNaN(" ")` returns false... Why? Because javascript casts the empty space`" "` as a `0`, then `isNaN()` is tricked. Another one could be `isNaN(true)` that is the equivalent to say `isNaN(1)`
For those reasons i found that the best way in this case in particular is using `isNaN()` but with an if before, checking if there is an empty space. Due to the nature of the app it would never be possible to evaluate booleans like `true` and `false` because they will come to the function as strings, that would be `"true"` and `"false"`

mjuan99 commented on src/controller/healthController.js:
```diff
@@ -0,0 +1,9 @@
+
+export const getHealthController = async (req, res) => {
+  try {
+    res.status(200).json({ data: "Health connected to database" });
```
getHealthController is not checking de database connection, it's always returning with status 200

Sly-Agustin replied:
I think it has been fixed now

mjuan99 commented on src/controller/students/createStudentController.js:
```diff
@@ -0,0 +1,17 @@
+    try {
+        const { body } = req;
+        const credentials = process.env.MONGODB_URL;
+        await mongoose.connect(credentials);
```
I think there's no need to connect to de database every time you want to create a new student

mjuan99 commented on src/controller/students/createStudentController.js:
```diff
@@ -0,0 +1,17 @@
+
+        const newStundent = new Student(body)
+        await newStundent.save();
+        res.status(200).json({ message: "Student created succesfully" });
```
There's no '400: Bad Request' response when creating a new student

mjuan99 commented on src/test/createQuestionnaire.test.js:
```diff
@@ -0,0 +1,151 @@
+        }
+      },
+    }
+    expect(await createQuestionnaire(Questionnaire, invalidObject)).toEqual({
```
This call is not mocked, it's using the real mongoose model

Sly-Agustin replied:
I have modified the test, check it out later

Sly-Agustin committed:
Fixes required by pull request #1

- Changed the way the id is parsed, this is due to inconsistiencies in how parseInt works internally. Also a special case with empty spaces is considered because javascript interpret it as a 0, so isNaN would fail to catch it.
- Health controller now checks the connection based on  the actual readyState
- Removed redundant connection on createStudentController

Sly-Agustin committed:
Fixed createQuestionnaire test

- Test is mocked correctly this time

Sly-Agustin committed:
Added student validation

- createStudentController now validates the data it receives to create a new one

Sly-Agustin committed:
Update README.md

Sly-Agustin committed:
Updated documentation

- Updated docs with the .nvmrc file

diazmaira commented on README.md:
```diff
@@ -1,2 +1,34 @@
 Repo for learning and test things with NoSQL
+
+# Pre-requisites
+Node and npm must be installed for installation and usage of the app. Node version used is 14.17.4
```
Nit: Take care of this line _"Node version used is 14.17.4"_. You keep the nodejs version in `.nvmrc` and `package.json`. If for some reason you update the nodejs version, you **must** update (by hand) the `README.md`. 
We as developers are always looking for automatize everything that is possible.

Sly-Agustin replied:
Added more generic explanation in https://github.com/Sly-Agustin/nosql-practice/commit/ff1e0aac702a96cdbcecc5e84ca69120ca8e1ae3

diazmaira commented on README.md:
```diff
@@ -1,2 +1,34 @@
+- Windows: `nvm use 14.17.4`, the windows version ignores the `.nvmrc` file completily, this is because is a different implementation of the GNU-Linux/MacOS one
+
+# Usage
+In your favourite terminal run `npm run start`, this initiates the app. In case you are running in a local environment usually the app will initiate in `127.0.0.1:3000`
```
The same as line `5`. If for some reason you update the public port, you  *must* update the port. For a more precise usage, you can add a Postman collection file.

Sly-Agustin replied:
Changed  the static version with the `env` and `.nvmrc` description in https://github.com/Sly-Agustin/nosql-practice/commit/ff1e0aac702a96cdbcecc5e84ca69120ca8e1ae3

diazmaira commented on README.md:
```diff
@@ -1,2 +1,34 @@
+# Usage
+In your favourite terminal run `npm run start`, this initiates the app. In case you are running in a local environment usually the app will initiate in `127.0.0.1:3000`
+
+Routes available are:
```
A better specification about endpoint is with the HTTP method. For example: `GET /health`.

Sly-Agustin replied:
Added the `GET` and `POST` for HTTP methods. Also added a brief description of the port and the `.env` file. All of this in https://github.com/Sly-Agustin/nosql-practice/commit/ff1e0aac702a96cdbcecc5e84ca69120ca8e1ae3

diazmaira commented on src/controller/createQuestionnaireController.js:
```diff
@@ -0,0 +1,45 @@
+function getErrorsFromValidation(error) {
+  let errorsMessage = {}
+  let keys = Object.keys(error.errors);
+  for(let i = 0; i<keys.length; i++) {
```
You can use  `map` function and  then variable  `keys` is no longer useful, but inside map scope.

Sly-Agustin replied:
Changed implementation to use a map in https://github.com/Sly-Agustin/nosql-practice/commit/c5b083ab932a40c68edab71d8db9198264581b81

diazmaira commented on src/controller/createQuestionnaireController.js:
```diff
@@ -0,0 +1,45 @@
+const createQuestionnaire = async (questionnaireModel, questionnaireData) => {
+  const questionnaire = new questionnaireModel(questionnaireData);
+  let error = questionnaire.validateSync();
+  if(error!=undefined) {
```
If you do not distinguish between `undefined` and `null`, then you should ask:

And you can also use conditional operator https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator

```
error ? getErrorsFromValidation(error) : questionarie;
```

Sly-Agustin replied:
Changed implementation to use ternary operator in https://github.com/Sly-Agustin/nosql-practice/commit/c5b083ab932a40c68edab71d8db9198264581b81

diazmaira commented on src/models/dream.js:
```diff
@@ -2,48 +2,102 @@ const mongoose = require('mongoose');
+            message   : '{VALUE} is not an integer value'
+        }
+    },
+    questionnaireId: {
```
`mongoose` creates an `id`. Use it, instead of implement yours.

Sly-Agustin replied:
Changed in https://github.com/Sly-Agustin/nosql-practice/commit/33e2cf59f39053cc2ae242441b93b751c2accdd0

diazmaira commented on src/controller/getDreamControllerById.js:
```diff
@@ -0,0 +1,49 @@
+
+async function getDreamControllerById (req, res) {
+    try{
+        const id = idVerification(req.params.id);
```
`id` should be a mongoose object id, which it is not a number, it is a 24-character decimal string, for example `5e1a0651741b255ddda996c4`. `mongoose` creates them and also it has its own validator. So let `mongoose` validates the params.id.

Sly-Agustin replied:
Changed in https://github.com/Sly-Agustin/nosql-practice/commit/33e2cf59f39053cc2ae242441b93b751c2accdd0

diazmaira commented on src/controller/healthController.js:
```diff
@@ -0,0 +1,14 @@
+const mongoose = require('mongoose');
+
+export const getHealthController = async (req, res) => {
+  try {
```
There is a function that return the connection states `mongoose.STATES[mongoose.connection.readyState]`. You don't need to use `try..catch` (we trust in mongoose function 😂) , just return the value

Sly-Agustin replied:
Fixed in https://github.com/Sly-Agustin/nosql-practice/commit/05ca9d525d22267bdf2581cddb38435f45f0388b

diazmaira commented on src/server/routes/questionnaire.route.js:
```diff
@@ -0,0 +1,11 @@
+import { getDreamControllerById } from '../../controller/getDreamControllerById';
+const router = express.Router();
+
+router.post('/create',  createQuestionnaireController);
```
When create a resource in database, there is no need to add the verb `create`. It is an standard. So you can define `router.post('/', ...);`

Sly-Agustin replied:
Updated in https://github.com/Sly-Agustin/nosql-practice/commit/32e06334cb14ef863c9d06941c53a4b8bee84c7b

diazmaira commented on src/server/routes/student.route.js:
```diff
@@ -0,0 +1,7 @@
+import { createStudentController } from '../../controller/students/createStudentController';
+const router = express.Router();
+
+router.post('/createStudent',  createStudentController);
```
the same. If you use the verb `post` it is implicit that you are creating an student.

Sly-Agustin replied:
Updated in https://github.com/Sly-Agustin/nosql-practice/commit/32e06334cb14ef863c9d06941c53a4b8bee84c7b

Sly-Agustin committed:
Readme update

- Updated readme to comply with #1 changes requested
- Added a postman collection file

Sly-Agustin committed:
Updated createQuestionnaireController

- Updated the controller to comply with the issues in #1

Sly-Agustin committed:
Update healthController.js

Sly-Agustin committed:
Updated routes

- Updated routes
- Updated the postman collection to reflect changes

Sly-Agustin committed:
Updated model and getbyId

- Updated model to use id generated by mongoose instead of the ones generated by the trigger in mongodb
- Updated controller to use objectIDs instead of custom IDs

Sly-Agustin committed:
Changed message

- Now the message does not say that the ID must be a number, now it says its invalid

diazmaira commented:
Well done! You can merge the code to the main branch
